src/event-stream/tokens-contract-handler.ts

```diff
@@ -0,0 +1,397 @@
+          },
+          error: 'uint128',
+        },
+      },
+    },
+  },
+  {
+    access: 'public',
+    args: [
+      { type: 'uint128', name: 'id' },
+      { type: 'principal', name: 'sender' },
+      { type: 'principal', name: 'recipient' },
+    ],
+    name: 'transfer',
+    outputs: {
+      type: {
+        response: {
+          ok: 'bool',
+          error: {
+            tuple: [
+              { type: { 'string-ascii': { length: 32 } }, name: 'kind' },
+              { type: 'uint128', name: 'code' },
+            ],
+          },
+        },
+      },
+    },
+  },
+];
+
+interface TokenMetadata {
+  name: string;
+  imageUrl: string;
+  description: string;
+}
+
+export class TokensContractHandler {
+  private contractAddress: string;
+  private contractName: string;
+  private contractAbi: ClarityAbi;
+  private db: DataStore;
+  private randomPrivKey = makeRandomPrivKey();
+  private chainId: ChainID;
+  private stacksNetwork: StacksNetwork;
+  private address: string;
+
+  constructor(
+    contractAddress: string,
+    contractName: string,
+    smartContractAbi: ClarityAbi,
+    datastore: DataStore,
+    chainId: ChainID
+  ) {
+    this.contractAddress = contractAddress;
+    this.contractName = contractName;
+    this.contractAbi = smartContractAbi;
+    this.db = datastore;
+    this.chainId = chainId;
+
+    this.stacksNetwork = GetStacksNetwork(this.chainId);
+    this.address = getAddressFromPrivateKey(
+      this.randomPrivKey.data,
+      this.chainId === ChainID.Mainnet ? TransactionVersion.Mainnet : TransactionVersion.Testnet
+    );
+  }
+  async start() {
+    if (this.contractAbi.fungible_tokens.length > 0) {
+      if (this.isCompliance(FT_FUNCTIONS)) {
+        await this.handleFTContract();
+      }
+    }
+
+    if (this.contractAbi.non_fungible_tokens.length > 0) {
+      if (this.isCompliance(NFT_FUNCTIONS)) {
+        await this.handleNFTContract();
+      }
+    }
+  }
+
+  /**
+   * fetch Fungible contract metadata
+   */
+  private async handleFTContract() {
+    const contractCallName = await this.makeReadOnlyContractCall('get-name', []);
+    const contractCallUri = await this.makeReadOnlyContractCall('get-token-uri', []);
+    const contractCallSymbol = await this.makeReadOnlyContractCall('get-symbol', []);
+    const contractCallDecimals = await this.makeReadOnlyContractCall('get-decimals', []);
+
+    const nameCV = this.checkAndParseString(contractCallName);
+    const uriCV = this.checkAndParseOptionalString(contractCallUri);
+    const symbolCV = this.checkAndParseString(contractCallSymbol);
+    const decimalsCV = this.checkAndParseUintCV(contractCallDecimals);
+
+    if (uriCV) {
+      const metadata = await this.getMetadataFromUri(uriCV.data);
+
+      const fungibleTokenMetadata: DbFungibleTokenMetadata = {
+        name: nameCV ? nameCV.data : metadata.name, //prefer the on-chain name
+        description: metadata.description,
+        image_uri: metadata.imageUrl,
```

Comment:
where does `imageUrl` comes from?  the json itself will include just `image` for the uri of the graphic file to use.

Reply:
I'm using this as a ref https://ipfs.io/ipfs/QmPAg1mjxcEQPPtqsLoEcauVedaeMH81WXDPvPx3VC5zUz. It has `imageURL` key.

Reply:
am I using the right value?

Reply:
see the FT sip schema (yeah, not so readable, but you can paste the schema into https://extendsclass.com/json-schema-validator.html to get an an example file that follows the schema), but here's the image key https://github.com/stacksgov/sips/pull/5/files#diff-b83747873c039e89a2dbcc593f0934ce59aaab9f58c8cacaefe4c01f6f2bdc95R143

some example files that follow the schema: https://swapr.finance/tokens/xusd.json, https://swapr.finance/tokens/token-soft-xbtc.json (you can ignore the `vector` key, but this just shows people can extend the original schema, and you can ignore these.

that's for FTs, not following the NFT SIP as closely, but there is a new SIP to describe this, so maybe not ready.  Others may have more precise information.

Reply:
It seems we have to treat FT and NFT token uri differently. It would be very helpful if anybody has concrete information regarding it.

Reply:
I'm only familiar with the FT side of things, so maybe @friedger could chime in?

Reply:
There is no specification for the meta data that should be returned by get-token-uri.
I usually recommend the meta data standard by open sea https://docs.opensea.io/docs/metadata-standards. Risidio are reviewing their current meta data schema.